system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -372,6 +393,396 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>
+        if (!location)
+            throw MakeStringException(XPATHERR_InvalidState,"XpathContext:addElementToLocation: failed to add element '%s'", name);
+        setLocation(location);
+    }
+
+    virtual void addXmlContent(const char *xml) override
+    {
+        xmlNodePtr location = m_xpathContext->node;
+        xmlParserCtxtPtr parserCtx = xmlCreateDocParserCtxt((const xmlChar *)xml);
+        if (!parserCtx)
+            throw MakeStringException(-1, "CLibXpathContext:addXmlContent: Unable to parse xml");
+        parserCtx->node = location;
+        xmlParseDocument(parserCtx);
+        int wellFormed = parserCtx->wellFormed;
+        xmlFreeDoc(parserCtx->myDoc); //dummy document
+        xmlFreeParserCtxt(parserCtx);
+        if (!wellFormed)
+            throw MakeStringException(-1, "XpathContext:addXmlContent: Unable to parse %s XML content", xml);
+
+    }
+    virtual bool ensureLocation(const char *xpath, bool required) override
+    {
+        if (isEmptyString(xpath))
+            return true;
+        if (strstr(xpath, "//")) //maybe in the future, scripting error rather than not found, required==false has no effect
+            throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: '//' not currently allowed '%s'", xpath);
+
+        xmlNodePtr current = m_xpathContext->node;
+        xmlNodePtr node = current;
+        if (*xpath=='/')
+        {
+            xpath++;
+            if (strncmp(xpath, "esdl_script_context/", 20)!=0)
+                throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: incorect first absolute path node '%s'", xpath);
+            xpath+=20;
+            node = xmlDocGetRootElement(m_xmlDoc);
+        }
+
+        StringArray xpathnodes;
+        xpathnodes.appendList(xpath, "/");
+        ForEachItemIn(i, xpathnodes)
+        {
+            const char *finger = xpathnodes.item(i);
+            if (isEmptyString(finger)) //shouldn't happen, scripting error rather than not found, required==false has no effect
+                throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: '//' not allowed '%s'", xpath);
+            if (*finger=='@')
+                throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: attribute cannot be selected '%s'", xpath);
+
+            xmlXPathObjectPtr obj = xmlXPathNodeEval(node, (const xmlChar*) finger, m_xpathContext);
+            xmlXPathSetContextNode(current, m_xpathContext); //restore
+
+            if (!obj)
+                throw MakeStringException(XPATHERR_InvalidInput,"XpathContext:ensureLocation: invalid xpath '%s' at node '%s'", xpath, finger);
+
+            if (obj->type!=xmlXPathObjectType::XPATH_NODESET || !obj->nodesetval || obj->nodesetval->nodeNr==0)
+            {
+                if (*finger=='.' || strpbrk(finger, "[()]*") || strstr(finger, "::")) //complex nodes must already exist
```

Comment:
Will be good to ensure that we document this constraint.

Reply:
Yes.  Once explained in the docs it would be pretty clear why this is necessary.  For example I can't add a node named a*b, but I could fine a node that matched that pattern.